lib/ocpp/v201/charge_point.cpp

```diff
@@ -563,6 +589,113 @@ AuthorizeResponse ChargePoint::validate_token(const IdToken id_token, const std:
         return response;
     }
 
+    // C07: Authorization using contract certificates
+    if (id_token.type == IdTokenEnum::eMAID) {
+        // Temporary variable that is set to true to avoid immediate response to allow the local auth list
+        // or auth cache to be tried
+        bool tryLocalAuthListOrCache = false;
+        bool forwardToCsms = false;
+
+        // If OCSP data is provided as argument, use it
+        if (this->websocket->is_connected() and ocsp_request_data.has_value()) {
+            EVLOG_info << "Online: Pass provided OCSP data to CSMS";
+            response = this->authorize_req(id_token, std::nullopt, ocsp_request_data);
+            forwardToCsms = true;
+        } else if (certificate.has_value()) {
+            // First try to validate the contract certificate locally
+            CertificateValidationResult localVerifyResult =
+                this->evse_security->verify_certificate(certificate.value().get(), ocpp::CaCertificateType::MO);
+            EVLOG_info << "Local contract validation result: " << localVerifyResult;
+
+            bool CentralContractValidationAllowed =
+                this->device_model
+                    ->get_optional_value<bool>(ControllerComponentVariables::CentralContractValidationAllowed)
+                    .value_or(true);
+            bool ContractValidationOffline =
+                this->device_model->get_optional_value<bool>(ControllerComponentVariables::ContractValidationOffline)
+                    .value_or(true);
+            bool LocalAuthorizeOffline =
+                this->device_model->get_optional_value<bool>(ControllerComponentVariables::LocalAuthorizeOffline)
+                    .value_or(true);
+
+            // C07.FR.01: When CS is online, it shall send an AuthorizeRequest
+            // C07.FR.02: The AuthorizeRequest shall at least contain the OCSP data
+            // TODO: local validation results are ignored if response is based only on OCSP data, is that acceptable?
```

Comment:
I am not sure about this. One could even argue that we should always generate the OCSP data in libocpp instead of allowing this data coming from the outside but I don't know if that is possible. Are there sources that only give us the OCSP data?

The way it is now a caller to this function can give a cert chain with different data then the OCSP contains so then we might need to check. In that case not even having OCSP as an argument would be just as quick.

Reply:
I implemented it this way to avoid breaking existing functionality. AFAIK in EVerest, it is possible that local chain validation and generation of OCSP data is done already by the ISO15118 module.